### PR TITLE
refactor `get_columns`

### PR DESCRIPTION
**Pull request by HuaHuaY**
related #9

### CODE REVIEW

File: sqlalchemy_risingwave/base.py [hunk 1]
```diff
@@ -73,15 +45,25 @@ def _get_server_version_info(self, conn):
...
     def get_table_names(self, conn, schema=None, **kw):
-        # TODO: use \dt instead. Oddly there seems some escape words issue...
+        # # We can't use these statement unitl we set `default_schema_name`.
+        # sql = (
+        #     "SELECT table_name FROM information_schema.tables WHERE "
+        #     "table_schema = :table_schema"
+        # )
+        # rows = conn.execute(
+        #     text(sql),
+        #     {"table_schema": schema or self.default_schema_name},
```
Comment (BowenXiao1999) [inline sqlalchemy_risingwave/base.py:55]: Remove this when merged

File: sqlalchemy_risingwave/base.py [hunk 2]
```diff
@@ -93,25 +75,37 @@ def get_columns(self, conn, table_name, schema=None, **kw):
...
             # returned type object.
-            m = re.match(r"^(\w+(?: \w+)*)(?:\(([0-9, ]*)\))?$", type_str)
-            if m is None:
-                warn("Could not parse type name '%s'" % type_str)
-                typ = sqltypes.NULLTYPE
-            else:
-                type_name, type_args = m.groups()
+            m = re.match(r"^struct<([a-z ]+)>$", type_str)
+            if m:
+                sub_type = m.group(1)
```
Comment (BowenXiao1999) [inline sqlalchemy_risingwave/base.py:80]: IIUC, the struct type can have multiple types inside it? But here seems we only process one

File: sqlalchemy_risingwave/base.py [hunk 3]
```diff
@@ -93,25 +75,37 @@ def get_columns(self, conn, table_name, schema=None, **kw):
...
+            else:
+                m = re.match(r"^([a-z ]+)((\[\])*)$", type_str)
+                if m:
+                    groups = m.groups()
+                    type_name = groups[0]
+                    try:
+                        type_class = _type_map[type_name]
+                        if len(groups) > 1:
+                            array_suffixs = re.findall(r"\[\]", groups[1])
+                            for i in range(0, len(array_suffixs)):
```
Comment (BowenXiao1999) [inline sqlalchemy_risingwave/base.py:97]: Write some comments for how the type str looks like? And IIRC seems like Array only support one type.

**Review by BowenXiao1999 — COMMENTED** [review]
What about add some tests on some types?

File: sqlalchemy_risingwave/base.py (hunk 2 above)
Comment (HuaHuaY) [reply sqlalchemy_risingwave/base.py:80]: Sorry, I forgot to add comma. Regular expression should be "struct<([a-z ,]+)>$"

File: sqlalchemy_risingwave/base.py (hunk 3 above)
Comment (HuaHuaY) [reply sqlalchemy_risingwave/base.py:97]: RisingWave supports multi-dimensional array so I write this. We can also refuse multi-dimensional array.

File: sqlalchemy_risingwave/base.py (hunk 2 above)
Comment (BowenXiao1999) [reply sqlalchemy_risingwave/base.py:70]: I mean, in a struct there may be mutiple types, but seems like we only process one type `sub_type`?

File: sqlalchemy_risingwave/base.py [hunk 4]
```diff
@@ -9,47 +9,19 @@
...
+    "integer": sqltypes.INT,  # DataType::Int32
+    "bigint": sqltypes.BIGINT,  # DataType::Int64
+    "real": sqltypes.FLOAT,  # DataType::Float32
+    "double precision": sqltypes.FLOAT,  # DataType::Float64
+    "numeric": sqltypes.DECIMAL,  # DataType::Decimal
+    "date": sqltypes.DATE,  # DataType::Date
+    "varchar": sqltypes.VARCHAR,  # DataType::Varchar
+    "time without time zone": sqltypes.Time,  # DataType::Time
+    "timestamp without time zone": sqltypes.TIMESTAMP,  # DataType::Timestamp
+    "timestamp with time zone": sqltypes.TIMESTAMP,  # DataType::Timestampz
```
Comment (BowenXiao1999) [inline sqlalchemy_risingwave/base.py:23]: So now for timestamp type, the string is `timestamp with time zone` or just `timestamp`?

File: sqlalchemy_risingwave/base.py [hunk 5]
```diff
@@ -9,47 +9,19 @@
...
-    "jsonb": sqltypes.JSON,
-    "uuid": UUID,
-    "inet": INET,
+    "boolean": sqltypes.BOOLEAN,  # DataType::Boolean
+    "smallint": sqltypes.SMALLINT,  # DataType::Int16
+    "integer": sqltypes.INT,  # DataType::Int32
+    "bigint": sqltypes.BIGINT,  # DataType::Int64
+    "real": sqltypes.FLOAT,  # DataType::Float32
+    "double precision": sqltypes.FLOAT,  # DataType::Float64
+    "numeric": sqltypes.DECIMAL,  # DataType::Decimal
```
Comment (BowenXiao1999) [inline sqlalchemy_risingwave/base.py:18]: Can we add some type specific info like the precision or etc for decimal?

File: sqlalchemy_risingwave/base.py (hunk 2 above)
Comment (HuaHuaY) [reply sqlalchemy_risingwave/base.py:70]: My mistake here. I will fix it later.

File: sqlalchemy_risingwave/base.py (hunk 4 above)
Comment (HuaHuaY) [reply sqlalchemy_risingwave/base.py:23]: Timestamp is "timestamp without time zone". It's the Display trait's output of struct `DataType`. You can check them in `src/common/src/types/mod.rs`.

File: sqlalchemy_risingwave/base.py (hunk 5 above)
Comment (HuaHuaY) [reply sqlalchemy_risingwave/base.py:18]: Now binder in RisingWave only allows `Decimal` without precision in function `bind_data_type`. And `DataType` in RisingWave doesn't record precision. But this grammar may occur in many third-party clients' test cases. I think we can create an issue in RisingWave and then support precision.

File: sqlalchemy_risingwave/base.py (hunk 2 above)
Comment (BowenXiao1999) [reply sqlalchemy_risingwave/base.py:70]: Yes. If it is non-trival to add `struct`, let's do it in other issues.

**Comment by BowenXiao1999** [comment]
Cool. The last step is to verifiy via tests.

Do you think there are some py tests we can add or integrate with Superset to check some type is worked correctly (new added `array`, `timestamp`, `int` etc, by create table and do some query): https://www.notion.so/risingwave-labs/Integration-with-Apache-Superset-e6468c23012045fea57d16a1f787e6b7

File: sqlalchemy_risingwave/base.py (hunk 2 above)
Comment (HuaHuaY) [reply sqlalchemy_risingwave/base.py:70]: I plan to change to this: If 'struct<' is at the beginning and '>' is at the ending, directly prompts that struct is not supported. Because nested structs are hard to parse.

Comment (BowenXiao1999) [reply sqlalchemy_risingwave/base.py:70]: IIUC, this is same to reject all sturtc type, right?

Comment (HuaHuaY) [reply sqlalchemy_risingwave/base.py]: Yes. `struct` does not have a corresponding type in sql. Here I just want to check the legitimacy of struct type. But I found it is too complex.
